src/parcels/_core/sgrid.py

```diff
@@ -0,0 +1,373 @@
+
+Note this code doesn't aim to completely cover the SGrid conventions, but aim to
+cover SGrid to the extent to which Parcels is concerned.
+"""
+
+from __future__ import annotations
+
+import enum
+import re
+from collections.abc import Hashable, Iterable
+from dataclasses import dataclass
+from typing import Any, Literal, Protocol, Self, overload
+
+import xarray as xr
+
+RE_DIM_DIM_PADDING = r"(\w+):(\w+)\s*\(padding:\s*(\w+)\)"
+
+Dim = str
+
+
+class Padding(enum.Enum):
+    NONE = "none"
+    LOW = "low"
+    HIGH = "high"
+    BOTH = "both"
+
+
+class SGridMetadataProtocol(Protocol):
+    def to_attrs(self) -> dict[str, str | int]: ...
+    def from_attrs(cls, d: dict[str, Hashable]) -> Self: ...
+
+
+class Grid2DMetadata(SGridMetadataProtocol):
+    def __init__(
+        self,
+        cf_role: Literal["grid_topology"],
+        topology_dimension: Literal[2],
+        node_dimensions: tuple[Dim, Dim],
+        face_dimensions: tuple[DimDimPadding, DimDimPadding],
+        vertical_dimensions: None | tuple[DimDimPadding] = None,
+    ):
+        if cf_role != "grid_topology":
+            raise ValueError(f"cf_role must be 'grid_topology', got {cf_role!r}")
+
+        if topology_dimension != 2:
+            raise ValueError("topology_dimension must be 2 for a 2D grid")
+
+        if not (
+            isinstance(node_dimensions, tuple)
+            and len(node_dimensions) == 2
+            and all(isinstance(nd, str) for nd in node_dimensions)
+        ):
+            raise ValueError("node_dimensions must be a tuple of 2 dimensions for a 2D grid")
+
+        if not (
+            isinstance(face_dimensions, tuple)
+            and len(face_dimensions) == 2
+            and all(isinstance(fd, DimDimPadding) for fd in face_dimensions)
+        ):
+            raise ValueError("face_dimensions must be a tuple of 2 DimDimPadding for a 2D grid")
+
+        if vertical_dimensions is not None:
+            if not (
+                isinstance(vertical_dimensions, tuple)
+                and len(vertical_dimensions) == 1
+                and isinstance(vertical_dimensions[0], DimDimPadding)
+            ):
+                raise ValueError("vertical_dimensions must be a tuple of 1 DimDimPadding for a 2D grid")
+
+        # Required attributes
+        self.cf_role = cf_role
+        self.topology_dimension = topology_dimension
+        self.node_dimensions = node_dimensions
+        self.face_dimensions = face_dimensions
+
+        #! Optional attributes aren't really important to Parcels, can be added later if needed
+        # Optional attributes
+        # # With defaults (set in init)
+        # edge1_dimensions: tuple[Dim, DimDimPadding]
+        # edge2_dimensions: tuple[DimDimPadding, Dim]
+
+        # # Without defaults
+        # node_coordinates: None | Any = None
+        # edge1_coordinates: None | Any = None
+        # edge2_coordinates: None | Any = None
+        # face_coordinate: None | Any = None
```

Comment:
Are these debug statements, or intentional comments? If the latter, is this the best place for them?